### PR TITLE
Support tracing native functional collective via python APIs

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119951
* #119950
* #119881
* #119229
* #119104
* __->__ #119103

Summary:
- Inlined `torch.distributed.distributed_c10d._get_group_size_by_name`
- Updated all torch.compile tests in test_c10d_functional_native.py to use funcol python APIs (as opposed to the dispatcher ops)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/distributed/test_c10d_functional_native.py [hunk 1]
```diff
@@ -499,10 +499,8 @@ def test_inductor_all_gather_into_tensor_single(self):
         self._init_process_group()
 
         def func(arg: torch.Tensor) -> torch.Tensor:
-            ag0 = torch.ops._c10d_functional.all_gather_into_tensor(
-                arg, self.world_size, "default"
-            )
-            ag0 = torch.ops._c10d_functional.wait_tensor(ag0)
+            ag0 = funcol.all_gather_tensor(arg, 0, "default")
```
Comment (wconstab) [inline test/distributed/test_c10d_functional_native.py:502]: what happened to the 'into', was that intentional?

**Review by wconstab — APPROVED** [review]
dynamo part and test changes lgtm, except one possible typo i commented (pls fix or explain before merge)

**Review by fegin — APPROVED** [review]
nice!

File: test/distributed/test_c10d_functional_native.py (hunk 1 above)
Comment (wanchaol) [reply test/distributed/test_c10d_functional_native.py:502]: iirc the python api is `funcol.all_gather_tensor`, and the ops binded to the c10d_functional namespace is sth like `all_gather_into_tensor`. Maybe we should make the naming consistent to `all_gather_tensor` in the future

**Review by wanchaol — APPROVED** [review]

**Pull request closed without merging**